# Projeto 1 - Ciência dos Dados

Nome: Breno Alencar Araújo

Nome: Celina Vieira de Melo

Atenção: Serão permitidos grupos de três pessoas, mas com uma rubrica mais exigente. Grupos deste tamanho precisarão fazer um questionário de avaliação de trabalho em equipe

___
Carregando algumas bibliotecas:

In [255]:
%matplotlib inline
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import os

In [256]:
print('Esperamos trabalhar no diretório')
print(os.getcwd())

Esperamos trabalhar no diretório
C:\Users\Breno Alencar\Documents\Insper\Semestre 2\CDados\proj2copia2


Carregando a base de dados com os tweets classificados como relevantes e não relevantes:

In [257]:
filename = 'Windows 10_final.xlsx'

No projeto 1 escolhemos a marca Windows 10 para fazer a análise dos Tweets para a empresa! Com isso, classificamos como relevante os Tweets que estavam relacionados as opiniões dos usuários em relação ao produto e como irrelevantes anúncios, links para correção de erros do Windows etc.

In [262]:
#Criando Dataframe da planilha Treinamento
train = pd.read_excel(filename)
train.head(5)

Treinamento  Relevância
0  relógio casio que eu não usava há meses com a ...           1
1  rt @pplware: hoje trazemos 3 ferramentas que o...           0
2  ct news - 15/03/2021 (windows 10 começa a dar ...           1
3  quer comprar este produto?entao clique neste l...           0
4  você ainda pode fazer upgrade gratuito para o ...           1

In [263]:
#Criando Dataframe da aba Teste da planilha
test = pd.read_excel(filename, sheet_name = 'Teste')
test.head(5)

Teste  Relevância
0   um dia eu restauro um windows 10 outro dia um xp           1
1  rt @ajulianafiuza: vivendo ou atualizando o wi...           1
2  finalmente voltei o meu pc p windows 10 obg se...           1
3     @doliveira__ esse aqui https://t.co/fpcbvgkkhj           1
4  todo dia esse windows 10 pedindo atualização, ...           1

In [327]:
#Pd Series de Treinamento
treinamento = pd.Series(train.Treinamento)


___
## Classificador automático de sentimento


O Windows 10 é o mais novo sistema operacional da Microsft e desde seu lançamento está rodeado de opiniões díspares.  
Para nosso classificador Naive-Bayes, consideramos como relevantes as constantes atualizações que o sistema passa, o tempo da atualização e o desempenho em jogos, além das experiências dos usuários do sistema. 

___
### Montando um Classificador Naive-Bayes

Considerando apenas as mensagens da planilha Treinamento, ensine  seu classificador.

In [459]:
import re 

def limpeza(text):
    """
        Função de limpeza que troca alguns sinais básicos por espaços
    """
    #importando string
    punctuation = '[!-.:?;''""@"|"\n"/]' # Sinais [] como delimitadores de um conjunto.
    pattern = re.compile(punctuation)
    text_subbed = re.sub(pattern, '', text)
    return text_subbed

In [460]:
#Removendo espaços dos Tweets
classifi_0 =' '.join(train.loc[filt_0,'Treinamento'])
classifi_0_limpo = limpeza(classifi_0.lower())

In [461]:
#Filtrando os Tweets Relevantes e Irrelevantes da aba Treinamento

filt_1 = train.Relevância == 1
classifi_1 =' '.join(train.loc[filt_1,'Treinamento'])


filt_0 = train.Relevância == 0
classifi_0 =' '.join(train.loc[filt_0,'Treinamento'])

#A função .lower() deixa todos os caracteres em minúsculo
classifi_1_limpo = limpeza(classifi_1.lower())


In [462]:
#Todas as palavras separadas dos tweets Relevantes e Irrelevantes
todas_os_tweets_irrelevantes = classifi_0_limpo.split()

todas_os_tweets_relevantes = classifi_1_limpo.split()
todas_os_tweets_relevantes
#print(len(todas_os_tweets_relevantes))


['relógio',
 'casio',
 'que',
 'eu',
 'não',
 'usava',
 'há',
 'meses',
 'com',
 'a',
 'hora',
 'certinha',
 'e',
 'windows',
 '10',
 'que',
 'eu',
 'uso',
 'todo',
 'santo',
 'dia',
 'atrasado',
 '2',
 'minutos',
 'bill',
 'gates',
 'vc',
 'perdeu',
 'ct',
 'news',
 '15032021',
 'windows',
 '10',
 'começa',
 'a',
 'dar',
 'tela',
 'azul',
 'com',
 'impressoras',
 'httpstco4ejai7pldm',
 'você',
 'ainda',
 'pode',
 'fazer',
 'upgrade',
 'gratuito',
 'para',
 'o',
 'windows',
 '10',
 'diz',
 'microsoft',
 'httpstcoqydzjjk9x8',
 'httpstcoy3ew0obnuy',
 'adinezinha',
 'eu',
 'pesquisei',
 'no',
 'google',
 'e',
 'a',
 'atualização',
 'é',
 'muitooo',
 'grande',
 'e',
 'tbm',
 'vi',
 'muita',
 'gente',
 'reclamando',
 'que',
 'tá',
 'há',
 'horas',
 'esperando',
 'o',
 'windows',
 'atualizar',
 'tbm',
 'é',
 'essa',
 'aquihttpstcozatwglzqzw',
 'carai',
 'todo',
 'dia',
 'atualização',
 'do',
 'windows',
 '10',
 'se',
 'fode',
 'microsoft',
 'rt',
 'olhardigital',
 'novo',
 'bug',
 'no',
 'wi

In [463]:
#Criando um pd.serie para todos os Tweets Irrelevantes
serie_tweets_irr = pd.Series(todas_os_tweets_irrelevantes)

#Criando um pd.serie para todos os Tweets Relevantes
serie_tweets_r = pd.Series(todas_os_tweets_relevantes)


In [464]:
#Criando Tabela de Frequência absoluta dos Tweets Irrelevantes
tabela_tweets_irr = serie_tweets_irr.value_counts()

#Criando Tabela de Frequência absoluta dos Tweets Relevantes
tabela_tweets_r = serie_tweets_r.value_counts()


In [465]:
tabela_tweets_relativa_irr = serie_tweets_irr.value_counts(True)

In [466]:
tabela_tweets_relativa_r = serie_tweets_r.value_counts(True)
#len(tabela_tweets_relativa_r)

In [467]:
tabela_tweets_relativa_r.sum()

0.9999999999999999

In [468]:
palavras_tweets = classifi_0_limpo + classifi_1_limpo
# len(palavras_tweets)

#Universo de palavras vindo dos tweets relevantes e irrelevantes da aba treinamento de nossa planilha
palvras_tweets = todas_os_tweets_irrelevantes + todas_os_tweets_relevantes

In [469]:
todas_os_tweets = palavras_tweets.split()
serie_tweets = pd.Series(todas_os_tweets)

In [470]:
# tabela_relativo_tweets = serie_tweets.value_counts(True)
tabela_absoluta_tweets = serie_tweets.value_counts()
tabela_absoluta_tweets

windows           291
10                256
o                 181
de                179
que               128
                 ... 
tarefas             1
r432264🛒acesse      1
x512jpej228t        1
desconto            1
execute             1
Length: 2111, dtype: int64

In [471]:
tabela_absoluta_tweets.shape[0]

2111

In [472]:
probabilidade_relevante = len(todas_os_tweets_relevantes)/len(todas_os_tweets)
probabilidade_relevante

0.6376153966515412

In [473]:
probilidade_irrelevantes = len(todas_os_tweets_irrelevantes)/len(todas_os_tweets)
probilidade_irrelevantes

0.36254107338444685

___
### Verificando a performance do Classificador

Agora você deve testar o seu classificador com a base de Testes.

In [474]:
quantidade_de_tweets = len(test['Teste'])

In [475]:
frase = (test['Teste'])


In [476]:
lista = list(frase)

lista_1 = []

j = 0
while j < len(lista):
    lista_1.append(lista[j])
    j+=1
    
frase = lista_1
print(frase)

['um dia eu restauro um windows 10 outro dia um xp', 'rt @ajulianafiuza: vivendo ou atualizando o windows 10?', 'finalmente voltei o meu pc p windows 10 obg senhor', '@doliveira__ esse aqui https://t.co/fpcbvgkkhj', 'todo dia esse windows 10 pedindo atualização, que droga\n por isso o 7 é o melhor.', 'a raiva que eu tô passando com essa atualização do windows 10 não é brincadeira', 'geralmente esse alguém é um desses 3 aqui:\n\neduardo vasconcelos \nbolsonaro \nwindows 10 https://t.co/7fuz4ogkha', 'merda tenho que att o windows 10 eu não quero essa merda', '@akkarin314 @juniorgalvaomvp @msdevbr hoje quase impossível instalar um pirata tipo eu mesmo o windows já reconhece a chave original na bios... windows 10 só tenho tido problemas com derrepende as thumbs do desktop demorarem a carregar.', '@pitodipaiagames @nerdlendario mas também não vejo que o xbox (console) está disputando com o pc, pois o windows 10 e xcloud hoje estão dentro do ecossistema xbox.', 'windows 10: atualização corri

In [477]:
lista_relevante = []
lista_irrelevante = []

for i in frase:
    i = i.split()
    for palavra in i:
        
        probFraseDadoR = 1
        
        if palavra in tabela_tweets_r:
            #j = todas_os_tweets_irrelevantes.index(palavra) 
            probFraseDadoR *= (tabela_tweets_r[palavra] + 1) / ( tabela_absoluta_tweets.shape[0] + len(todas_os_tweets_relevantes) )
            
        else:
            probFraseDadoR *= (0 + 1) / ( tabela_absoluta_tweets.shape[0] + len(todas_os_tweets_relevantes) )
            
                
    lista_relevante.append(probFraseDadoR)
                
for k in frase:
    k = k.split()
    for palavra in k:
        
        probFraseDadoIRR = 1
        
        if palavra in tabela_tweets_irr:
            #k = todas_os_tweets_relevantes.index(palavra)
            probFraseDadoIRR *= (tabela_tweets_irr[palavra] + 1) / ( tabela_absoluta_tweets.shape[0] + len(todas_os_tweets_irrelevantes) )
#            
          
        else: 
            probFraseDadoIRR *= (0 + 1) / ( tabela_absoluta_tweets.shape[0] + len(todas_os_tweets_irrelevantes) )
#             
        
    lista_irrelevante.append(probFraseDadoIRR)

In [478]:
print(len(lista_relevante))

208


In [479]:
print(len(lista_irrelevante))

208


In [480]:
test

Teste  Relevância  \
0     um dia eu restauro um windows 10 outro dia um xp           1   
1    rt @ajulianafiuza: vivendo ou atualizando o wi...           1   
2    finalmente voltei o meu pc p windows 10 obg se...           1   
3       @doliveira__ esse aqui https://t.co/fpcbvgkkhj           1   
4    todo dia esse windows 10 pedindo atualização, ...           1   
..                                                 ...         ...   
203  olha, ter um computador com windows 10 sem ssd...           1   
204  todo o meu trabalho naquele ssd e windows 10 i...           1   
205  anotar pra não esquecer:\nnunca mais tentar in...           1   
206  @guilhox pois é, o windows 10 mesmo eu uso obr...           1   
207  você pode ter o pc com os melhores specs do mu...           1   

     Classificador  
0                1  
1                1  
2                1  
3                1  
4                1  
..             ...  
203              1  
204              1  
205              1  
206              1  
207              1  

[208 rows x 3 columns]

In [481]:
probabilidade_relevante = len(todas_os_tweets_relevantes)/len(todas_os_tweets)
probabilidade_irrelevante = len(todas_os_tweets_irrelevantes)/len(todas_os_tweets)
lista_classificador = []

contador = 0
while contador < len(lista_relevante):
    
    Prob_de_ser_Relevante = probabilidade_relevante*lista_relevante[contador]
    Prob_de_ser_Irrelevante = probabilidade_irrelevante*lista_irrelevante[contador]
    
    if Prob_de_ser_Relevante > Prob_de_ser_Irrelevante:
        lista_classificador.append(1)
    
    else:
        lista_classificador.append(0)
        
    contador +=1
        

test["Classificador"] = lista_classificador
test.head()

Teste  Relevância  \
0   um dia eu restauro um windows 10 outro dia um xp           1   
1  rt @ajulianafiuza: vivendo ou atualizando o wi...           1   
2  finalmente voltei o meu pc p windows 10 obg se...           1   
3     @doliveira__ esse aqui https://t.co/fpcbvgkkhj           1   
4  todo dia esse windows 10 pedindo atualização, ...           1   

   Classificador  
0              1  
1              1  
2              1  
3              1  
4              1

In [482]:
pd.crosstab(test["Relevância"],test["Classificador"],normalize=True)*100

Classificador         0          1
Relevância                        
0              3.846154  32.692308
1              0.000000  63.461538

In [483]:
tabela_tweets_relativa_r

windows         0.048834
10              0.043436
o               0.032393
de              0.026503
que             0.023313
                  ...   
retornar        0.000245
fznd            0.000245
impressoras…    0.000245
positiva        0.000245
paz             0.000245
Length: 1358, dtype: float64

In [456]:
len(tabela_tweets_irr)

1069

In [457]:
len(todas_os_tweets_irrelevantes)

2318

___
### Concluindo

A partir da tabela CrosTab concluímos um acerto de 63,46% dos casos em que marcamos como relevante e um acerto de 3,36% dos casos em que marcamos como irrelevante

Verdadeiro positivo: 63,46%
Falso positivo: 32.69%
Falso negativo: 0.0%
Verdadeiro negativo: 3.84%

Mais de um terço dos tweets foi classificado como falso positivo devido a fatores como ironia e ambiguidade. Vide exemplo: "nem a existência do windows 10 foi uma ideia tão idiota assim ". 

___
### Qualidade do Classificador a partir de novas separações dos tweets entre Treinamento e Teste

Caso for fazer esse item do Projeto

In [486]:
precisao_percent = []
train2 = pd.read_excel(filename)
test2 = pd.read_excel(filename, sheet_name='Teste')
train2 = train2.rename(columns={'Treinamento':'Todos'})
test2 = test2.rename(columns={'Teste':'Todos'})
Total = pd.concat([test2,train2])
Total

Todos  Relevância
0     um dia eu restauro um windows 10 outro dia um xp           1
1    rt @ajulianafiuza: vivendo ou atualizando o wi...           1
2    finalmente voltei o meu pc p windows 10 obg se...           1
3       @doliveira__ esse aqui https://t.co/fpcbvgkkhj           1
4    todo dia esse windows 10 pedindo atualização, ...           1
..                                                 ...         ...
295  @microsoftbr @microsoftajuda não consigo ajuda...           1
296  você pode ter o pc com os melhores specs do mu...           1
297  @clapiclapiclapi esse daqui é um i5, 8gb de ra...           0
298  windows 10 recebe correção temporária da micro...           1
299  rt @magazinewmti: notebook hp 250 g7 intel cor...           0

[508 rows x 2 columns]

In [ ]:
for i in range(100):
    

___
## Aperfeiçoamento:

Os trabalhos vão evoluir em conceito dependendo da quantidade de itens avançados:

* Limpar: \n, :, ", ', (, ), etc SEM remover emojis
* Corrigir separação de espaços entre palavras e emojis ou entre emojis e emojis
* Propor outras limpezas e transformações que não afetem a qualidade da informação ou classificação
* Criar categorias intermediárias de relevância baseadas na probabilidade: ex.: muito relevante, relevante, neutro, irrelevante, muito irrelevante (3 categorias: C, mais categorias conta para B)
* Explicar por que não posso usar o próprio classificador para gerar mais amostras de treinamento
* Propor diferentes cenários para Naïve Bayes fora do contexto do projeto
* Sugerir e explicar melhorias reais com indicações concretas de como implementar (indicar como fazer e indicar material de pesquisa)


___
## Referências

[Naive Bayes and Text Classification](https://arxiv.org/pdf/1410.5329.pdf)  **Mais completo**

[A practical explanation of a Naive Bayes Classifier](https://monkeylearn.com/blog/practical-explanation-naive-bayes-classifier/) **Mais simples**